File: Python practice/DP/maxProfit.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Aug  7 22:45:22 2020

@author: akashkumar
"""

def maxProfit(prices):
    if len(prices)==0:
        return 0
    buyprice=prices[0]
    profit=[0]*len(prices)
    for i in range(1,len(prices)):
        buyprice=min(buyprice,prices[i])
        maxprofit=max(profit[i-1],prices[i]-buyprice)
        profit[i]=maxprofit
    return profit[-1]
maxProfit([7,1,5,3,6,4])

File: Python practice/DP/maxProfit.java
class Solution {
    public int maxProfit(int[] prices) {
        int min_price = Integer.MAX_VALUE;
        int maxprofit = 0;
        for(int i = 0; i< prices.length; i++){
            if(prices[i] < min_price)
                min_price = prices[i];
            else if (prices[i] - min_price > maxprofit)
                maxprofit = prices[i] - min_price;
        }
        return maxprofit;
    }
}